docs/HTTP_SPEC.md

```diff
@@ -0,0 +1,104 @@
+# Lassie: HTTP Specification
+
+![wip](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square)
+
+**Author(s)**:
+
+- [Hannah Howard](https://github.com/hannahhoward)
+- [Kyle Huntsman](https://github.com/kylehuntsman)
+- [Rod Vagg](https://github.com/rvagg)
+
+**Maintainer(s)**:
+
+- [Hannah Howard](https://github.com/hannahhoward)
+- [Kyle Huntsman](https://github.com/kylehuntsman)
+- [Rod Vagg](https://github.com/rvagg)
+
+* * *
+
+## Table of Contents
+
+- [Introduction](#introduction)
+- [Specification](#specification)
+    - [`GET /ipfs/{cid}[?params]`](#get-ipfscidparams)
+
+
+## Introduction
+
+The Lassie HTTP Daemon is an HTTP interface for retrieving CAR files from the Filecoin network. It fetches content via CIDs over a number of possible protocols, choosing the best protocol for the job.
+
+## Specification
+
+### `GET /ipfs/{cid}[?params]`
+
+Retrieves from peers that have the given CID content, downloading as a CAR file.
+
+#### Request
+
+##### Headers
+
+- `Accept` - _Optional_. Used to specify the response content type. Optional only if a `format` query parameter is provided, otherwise required.
+
+    If provided, the value must explicitly or implicitly include `application/vnd.ipld.car`.
+
+- `X-Request-Id` - _Optional_. Used to provide a unique request value that can be correlated with a unique retrieval ID in the logs.
+
+##### Path Parameters
+
+- `cid` - _Required_. The default string representation of the CID.
+
+##### Query Parameters
+
+- `filename` - _Optional_. Used to override the name of the downloaded CAR file.
+
+    The filename extension cannot be missing and must be `.car`.
+
+- `format` - _Optional_. `format=<format>` can be used to specify the response content type. This is a URL-friendly alternative to providing an `Accept` header. Optional only if an `Accept` header value is provided, otherwise required.
+
+    If provided, the format value must be `car`. Example: `format=car`.
+
+    `format=car` &rarr; `Accept: application/vnd.ipld.car`
+
+#### Response
+
+#### Status Codes
+
+- `200` - OK
+
+- `400` - Bad Request
+    - No acceptable content type provided in the `Accept` header
+    - Requested a non-supported format via the `format` query parameter
+    - Neither providing a valid `Accept` header or `format` query parameter
+    - No extension given in `filename` query parameter
+    - Used a non-supported extension in the `filename` query parameter
+
+- `404` - No candidates for the given CID were found
+
+- `500` - Internal Server Error
+    - The given CID path parameter could not be parsed
+    - An internal retrieval ID failed to generate
+    - The CAR file failed to write
```

Comment:
```suggestion
    - The internal blockstore file failed to write
```

Reply:
for user purposes, this failure is just from the initial, internal-only blockstore, the fact that it's a CAR is inconsequential, what the user gets is a separate thing.

also it sucks that we even have to deal with this thing, but we'll find a way to get rid of it eventually